python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,204 @@
+            Table: the updated table instance with its metadata
+
+        Raises:
+            TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def create_namespace(self, namespace: Union[str, Identifier], properties: Optional[Metadata] = None) -> None:
+        """Create a namespace in the catalog.
+
+        Args:
+            namespace: The namespace to be created. If the namespace is a string, it is split into a tuple on '.'. If it is a tuple, it is used as-is.
+            properties: A string dictionary of properties for the given namespace
+
+        Raises:
+            AlreadyExistsError: If a namespace with the given name already exists
+        """
+
+    @abstractmethod
+    def drop_namespace(self, namespace: Union[str, Identifier]) -> None:
+        """Drop a namespace.
+
+        Args:
+            namespace: The namespace to be dropped. If the namespace is a string, it is split into a tuple on '.'. If it is a tuple, it is used as-is.
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the given name does not exist
+            NamespaceNotEmptyError: If the namespace is not empty
+        """
+
+    @abstractmethod
+    def list_tables(self, namespace: Optional[Union[str, Identifier]] = None) -> List[Identifier]:
+        """List tables under the given namespace in the catalog.
+
+        If namespace not provided, will list all tables in the catalog.
+
+        Args:
+            namespace: the namespace to search. If the namespace is a string, it is split into a tuple on '.'. If it is a tuple, it is used as-is.
+
+        Returns:
+            List[Tuple[str, str]]: list of tuple of table namespace and their names.
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the given name does not exist
+        """
+
+    @abstractmethod
+    def list_namespaces(self) -> List[Identifier]:
+        """List namespaces from the given namespace. If not given, list top-level namespaces from the catalog.
+
+        Returns:
+            List[Identifier]: a List of namespace, where each element is a Tuple of namespace levels. Ex: ('com','org','dept')
+        """
+
+    @abstractmethod
+    def load_namespace(self, namespace: Union[str, Identifier]) -> Metadata:
+        """Get metadata dictionary for a namespace.
+
+        Args:
+            namespace: If the namespace is a string, it is split into a tuple on '.'. If it is a tuple, it is used as-is.
+
+        Returns:
+            Metadata: a dictionary of properties for the given namespace
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the given name does not exist
+        """
+
+    @abstractmethod
+    def update_namespace_metadata(
+        self, namespace: Union[str, Identifier], removals: Optional[Set[str]] = None, updates: Optional[Metadata] = None
+    ) -> None:
+        """Removes provided metadata keys and then updates metadata for a namespace.
```

Comment:
This should not specify an order of operations. It is an error if `removals` and `updates` have overlapping keys. There is no required ordering behavior.